python/pyiceberg/expressions/base.py

```diff
@@ -672,3 +683,166 @@ def visit_unbound_predicate(self, predicate) -> BooleanExpression:
+        """Visit a bound NotNan predicate"""
+
+    @abstractmethod
+    def visit_is_null(self, term: BoundTerm[T]) -> T:
+        """Visit a bound IsNull predicate"""
+
+    @abstractmethod
+    def visit_not_null(self, term: BoundTerm[T]) -> T:
+        """Visit a bound NotNull predicate"""
+
+    @abstractmethod
+    def visit_equal(self, term: BoundTerm[T], literal: Literal[Any]) -> T:
+        """Visit a bound Equal predicate"""
+
+    @abstractmethod
+    def visit_not_equal(self, term: BoundTerm[T], literal: Literal[Any]) -> T:
+        """Visit a bound NotEqual predicate"""
+
+    @abstractmethod
+    def visit_greater_than_or_equal(self, term: BoundTerm[T], literal: Literal[Any]) -> T:
+        """Visit a bound GreaterThanOrEqual predicate"""
+
+    @abstractmethod
+    def visit_greater_than(self, term: BoundTerm[T], literal: Literal[Any]) -> T:
+        """Visit a bound GreaterThan predicate"""
+
+    @abstractmethod
+    def visit_less_than(self, term: BoundTerm[T], literal: Literal[Any]) -> T:
+        """Visit a bound LessThan predicate"""
+
+    @abstractmethod
+    def visit_less_than_or_equal(self, term: BoundTerm[T], literal: Literal[Any]) -> T:
+        """Visit a bound LessThanOrEqual predicate"""
+
+    @abstractmethod
+    def visit_true(self) -> T:
+        """Visit a bound True predicate"""
+
+    @abstractmethod
+    def visit_false(self) -> T:
+        """Visit a bound False predicate"""
+
+    @abstractmethod
+    def visit_not(self, child_result: T) -> T:
+        """Visit a bound Not predicate"""
+
+    @abstractmethod
+    def visit_and(self, left_result: T, right_result: T) -> T:
+        """Visit a bound And predicate"""
+
+    @abstractmethod
+    def visit_or(self, left_result: T, right_result: T) -> T:
+        """Visit a bound Or predicate"""
+
+    def visit_unbound_predicate(self, predicate: UnboundPredicate[T]):
+        """Visit an unbound predicate
+        Args:
+            predicate (UnboundPredicate[T]): An unbound predicate
+        Raises:
+            TypeError: This always raises since an unbound predicate is not expected in a bound boolean expression
+        """
+        raise TypeError(f"Not a bound predicate: {predicate}")
+
+    def visit_bound_predicate(self, predicate: BoundPredicate[T]) -> T:
+        """Visit a bound predicate
+        Args:
+            predicate (BoundPredicate[T]): A bound predicate
+        """
+        return visit_bound_predicate(predicate, self)
+
+
+@singledispatch
+def visit_bound_predicate(obj, visitor: BooleanExpressionVisitor[T]) -> T:  # pylint: disable=unused-argument
+    pass
+
+
+@visit_bound_predicate.register(BoundIn)
+def _(obj: BoundIn, visitor: BoundBooleanExpressionVisitor[T]) -> T:
+    """Visit a bound In predicate"""
+    return visitor.visit_in(term=obj.term, literals=obj.literals)
+
+
+@visit_bound_predicate.register(BoundNotIn)
+def _(obj: BoundNotIn, visitor: BoundBooleanExpressionVisitor[T]) -> T:
```

Comment:
Nit: I'd probably rename `obj` to `expr` since it is an expression.

Reply:
Updated 👍